### ISSUE TITLE
Token desyncs with many shared services

### ISSUE THREAD

**Issue by Gaegamel** [privacyidea/privacyidea#428]
Hi,

**Infrastructure - more or less**
I would like to make use of privacyIDEA on my personal setup. It consists of two (three with privacyIDEA) services. Both support HOTP.

In order to separate both services and their OTPs from one another (I am thinking about opening one of the services to the public or at least a closed community) and to make management easier, I created two realms - realm A and realm B.

**Problem**
I want to use both services with my personal HOTP key, so I can login to both realms without needing a second key. But this arose a problem: I had to register two separate IDs for one token. They both have separate counters, which do not sync, because they do not know each other. Since I log in more frequently to realm A, the token will eventually lose sync with realm B. This happens from time to time.

**Question**
Is there any way (separating the realms for improved security) in order to sync all token-IDs, which have the same underlying shared secret?

**Remarks**
I suppose, greater scale organisations and setups might also profit from this solution, if each employee is only assigned one key and it's not required to possess one key for each service. of course, this could be toggled by a guideline or alike.

If you have further questions or need clarification, I would be glad to help you.

**Comment by cornelinux** [comment]
Your design is not right. Plus: This is the feature of the HOTP token. And it is always good to not mess with basic crypto functions.

You would not create one realm per service but have the users in one single realm and authenticate both services against this.
One user. One token.

If you still insist on using two realms, you can use the remote token and forward the request to the local machine to the realm token.

**Comment by Gaegamel** [comment]
Well, your first solution would work out. At least until the following scenario occurs:

1. User x and y are created. x is only granted access (via username+password) to service A and y only (via username+pw) to service B. Both services are based on an Apache 2.4 (and thus the privacyidea plugin with Basic Auth applies)
2. Both x and y receive a hardware-token. 
3. x was (illegally) able to get the credentials of y.

So far we could assume, x is still prohibited to access service B, since (by definition of 2FA) is not able to provide the second factor (the "possession") to the service. However, since x is allowed to use his own OTP (because they are checked against the same realm - which only answers the question "Are you a member of the same club?") he is able to bypass this security measurement.

So, basically two separate realms (with separate allowed tokens) are needed. And in order to keep the number of tokens limited, I was asking the first question. I suppose, this could also be considered an issue regarding the Apache plugin.

And regarding your second solution: Isn't this, in essence, causing the same problem like before? If the request is forwarded to this special realm, while the token of a second realm is not synchronized as well? One realm is constantly used, while the other runs out of sync.

**Comment by cornelinux** [comment]
You are messing up scenarios. One users or two users.
You may also want to take a look at the authorization policies.
If you need any further consulting on any probably complicated setup, please consider getting some consulting.
https://netknights.it/en/leistungen/one-time-services/
